fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/properties/CardinalitiesProperty.java

```diff
@@ -408,6 +410,16 @@ public Cardinalities visitRecordQueryInComparandJoinPlan(@Nonnull final RecordQu
         return Cardinalities.unknownMaxCardinality();
     }
 
+    @Nonnull
+    @Override
+    public Cardinalities visitRecordQueryRecursiveUnionPlan(@Nonnull final RecordQueryRecursiveUnionPlan element) {
+        final var initialStateCardinality = fromChild(element.getChildren().get(0));
+        // this can be improved by imposing an assertion on the cardinality of the recursive leg; if the recursive leg
+        // has a known minimum cardinality of x | x > 0 then this query is infinitely recursive, so we should probably
+        // throw.
```

Comment:
I'm not sure about that. For an infinite relation to be an error / illegal, you need the _whole_ context, and that's hard. For instance, isn't the ability to execute queries with a per-execution window size or the possibility of some streaming connection API that the user can close enough to say that sometimes you might really want it, even though it's impossible to materialize all of it?